### ISSUE TITLE
RMIA / LiRA ROC AUC far below paper (CIFAR-10, Wide-ResNet, 254 reference models)

### ISSUE THREAD

**Issue by yundaqwe** [privacytrustlab/ml_privacy_meter#146]
Dear authors, 
First, thanks for open-sourcing such a clean implementation—it worked smoothly out-of-the-box.

However, the MIA results I obtain are noticeably weaker than those reported in the paper.

# Results

| Parameter | Paper | Reproduced |
|----------|----------|----------|
| Avg. Train Accuracy | 99.9%  | 99.96%  |
| Avg. test Accuracy  | 92.4%  | 91.45 %  |
|AUC of RMIA (online)| 72.25 | 60.9|
|AUC of LiRA (online)| 72.04 | 61.1 |
|0.01 %TPR@FPR of LiRA (online)| 3.39%|0.31%|
|0.01 %TPR@FPR of RMIA (online)| 4.31%|0.18%|
# Environment
    "OS": "Linux-6.8.0-57-generic-x86_64-with-glibc2.35",
    "Python": "3.9.16 (main, Jan 11 2023, 16:05:54)  [GCC 11.2.0]",
    "CUDA": "Cuda compilation tools, release 11.6, V11.6.124",
    "GPU": "NVIDIA GeForce RTX 3090, 550.120\nNVIDIA GeForce RTX 3090, 550.120\nNVIDIA GeForce RTX 3090, 550.120\nNVIDIA GeForce RTX 3090, 550.120\nNVIDIA GeForce RTX 3090, 550.120\nNVIDIA GeForce RTX 3090, 550.120\nNVIDIA GeForce RTX 3090, 550.120\nNVIDIA GeForce RTX 3090, 550.120",
    "TensorFlow": "2.16.1",
    "PyTorch": "2.6.0+cu124",
    "TensorBoard": "2.16.2"
# Steps to Reproduce

1. bash train_cifar10.sh
2. bash infer_cifar10.sh
3. python main.py --cf attack_configs/cifar10/rmia_online_254_ref_models.yaml
4. python main.py --cf attack_configs/cifar10/lira_online_254_ref_models.yaml
# What I tried to resolve this issue
Verified training quality:  retrained reference models with test accuracy lower than 90%.
# Possible Cause
My best guess is a TensorFlow operator/numerical change affecting logits or gradient extraction for signals.
Happy to help bisect TF versions or run with a pinned Docker image if you have one.

Thanks again for your work again.
Best regards,
Zebin

**Comment by taojiashu** [comment]
First of all, thank you for trying our tool. But I am not sure how LiRA is run, as Privacy Meter does not implement it. So I will only help with the RMIA part. For RMIA, your numbers are really low. If you refer to the [sample output](https://github.com/privacytrustlab/ml_privacy_meter/blob/master/documentation/mia.md#auditing-results) we had for running the demo cifar-10 [config](https://github.com/privacytrustlab/ml_privacy_meter/blob/master/configs/cifar10.yaml), we have AUC over 0.81. 

In your steps to reproduce, I can see you have customized training scripts. It is quite difficult for me to figure out what is wrong with your customized scripts and attack configurations. Maybe you can attach the code here so that I can take a look?

At last, to reproduce the results in the RMIA paper, the best place to go is the research [folder](https://github.com/privacytrustlab/ml_privacy_meter/tree/d32734161a3395211fe5f3cd461932290b1fafbe/research/2024_rmia) in this repo. The RMIA and LiRA papers use customized WideResNets that use quite a lot of JAX tricks, which we cannot reproduce with PyTorch alone. Hence, I believe you have trained the JAX models and saved the inferred signals (in train_cifar10.sh and infer_cifar10.sh) to disk for the downstream membership inference (with Privacy Meter). But this process is not trivial and may incur errors too.

**Comment by yundaqwe** [comment]
Thank you for the quick follow-up. Just to reiterate: every result I reported in the previous issue was obtained by running the repository exactly as-is—no edits, no parameter tweaks, and no environment overrides. I firstly  I cloned the repo, checked out  this branch (https://github.com/privacytrustlab/ml_privacy_meter/tree/d32734161a3395211fe5f3cd461932290b1fafbe/research/2024_rmia) and then ran the code exactly as described in the [README](https://github.com/privacytrustlab/ml_privacy_meter/blob/d32734161a3395211fe5f3cd461932290b1fafbe/research/2024_rmia/README.md), without making any changes.  What I did is as follows:
1. bash [train_cifar10.sh ](https://github.com/privacytrustlab/ml_privacy_meter/blob/d32734161a3395211fe5f3cd461932290b1fafbe/research/2024_rmia/scripts/train_cifar10.sh)
2. bash [infer_cifar10.sh](https://github.com/privacytrustlab/ml_privacy_meter/blob/d32734161a3395211fe5f3cd461932290b1fafbe/research/2024_rmia/scripts/infer_cifar10.sh)
3. python [main.py](https://github.com/privacytrustlab/ml_privacy_meter/blob/d32734161a3395211fe5f3cd461932290b1fafbe/research/2024_rmia/main.py) --cf attack_configs/cifar10/[rmia_online_254_ref_models.yaml](https://github.com/privacytrustlab/ml_privacy_meter/blob/d32734161a3395211fe5f3cd461932290b1fafbe/research/2024_rmia/attack_configs/cifar10/rmia_online_254_ref_models.yaml)
4. python [main.py](https://github.com/privacytrustlab/ml_privacy_meter/blob/d32734161a3395211fe5f3cd461932290b1fafbe/research/2024_rmia/main.py) --cf attack_configs/cifar10/[lira_online_254_ref_models.yaml](https://github.com/privacytrustlab/ml_privacy_meter/blob/d32734161a3395211fe5f3cd461932290b1fafbe/research/2024_rmia/attack_configs/cifar10/lira_online_254_ref_models.yaml)

Thank you again for your assistance.

Best regards,
Zebin

**Comment by taojiashu** [comment]
This repo is for the open-source Python Library, ML Privacy Meter. If I understand correctly, you have an issue with the RMIA source code hosted in the research directory. Unfortunately, I cannot assist you with that because it is out of scope for our Privacy Meter repo, and we are not the authors of the source code.

Having said that, many of us had success reproducing the result with the source code. What I think is problematic in your steps is that you are using a mismatched attack configuration. The `train_cifar10.sh` and `infer_cifar10.sh` only contain code to work with up to 8 reference models, but you are using attack configs for 254 reference models. This can cause expected behavior in code execution and lead to unreliable results. I suggest using another attack config that corresponds to the number of reference models you have trained, or try [`attack_cifar10.sh`](https://github.com/privacytrustlab/ml_privacy_meter/blob/d32734161a3395211fe5f3cd461932290b1fafbe/research/2024_rmia/attack_cifar10.sh) instead of doing `python main.py`.

**Comment by yundaqwe** [comment]
Thank you very much for the clarification and your help!

Best
Zebin
